### PR TITLE
[stable/spinnaker] Allow serviceAccount annotations for Halyard 

### LINKED ISSUE #20885: [stable/spinnaker] Support for custom ServiceAccounts for Spinnaker Pods
**Is your feature request related to a problem? Please describe.**

TL;DR - Is it possible to update the ServiceAccount used by Spinnaker pods to be custom / modified / specific to the created Spinnaker pods, rather than the `default` ServiceAccount?

We wish to add IAM annotations to the ServiceAccount used by Spinnaker services, but do not want to give *any* pod using the `default` ServiceAccount the same permissions in AWS as Spinnaker

Currently a custom ServiceAccount can be used for the Halyard pod, but not the created Spinnaker pods (CloudDriver and Front50)

There is a note in [the RBAC section](https://github.com/helm/charts/blob/master/stable/spinnaker/templates/rbac/spinnaker-sa.yaml#L22-L23) that CloudDriver does not allow personal ServiceAccounts, but this was written [2 years ago](https://github.com/helm/charts/blame/master/stable/spinnaker/templates/rbac/spinnaker-sa.yaml)

(more context below)

**Describe the solution you'd like**
1. Set the ServiceAccount used by Spinnaker to be Spinnaker specific (like Halyard) not the namespace `default`
2. Any of the following:
  - Allow the passing of custom ServiceAccounts to Spinnaker pods
  - Allow annotations to be added to the ServiceAccount used by the Spinnaker pods (this would *require* the ServiceAccount used by Spinnaker to not be the `default` account)

I am happy to do a PR for this myself - creating the issue for visibility.
I am really just not sure if the RBAC note I mentioned above is still true - I was hoping someone from Spinnaker maintainers might have some insight

**Describe alternatives you've considered**
We currently use kube2IAM, but we would like to migrate away from this

**Additional context**
AWS have released [IAM Roles for ServiceAccounts](https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/) (IRSA) as a replacement / alternative for solutions like kube2IAM and others (in Sept 2019)

This allows IAM roles to be attached to specific ServiceAccounts, rather than as pod annotations - as the authentication is now on the pod itself, rather than the worker node having the permissions of all the pods in the cluster, this ties in with the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
There is also no race conditions etc as it is a webhook triggered on pod-creation.

Many Helm charts allow the passing of custom ServiceAccount names for the pods, or extra annotations to the created ServiceAccounts, allowing the rollout of IRSA:  [externalDNS](https://github.com/helm/charts/blob/master/stable/external-dns/templates/serviceaccount.yaml); [ALBIngressController](https://github.com/helm/charts/blob/master/incubator/aws-alb-ingress-controller/templates/serviceaccount.yaml)

In our clusters we have updated all our Helm services to use IRSA - apart from Spinnaker.

(EDIT 1 & 2: Formatting)

### PR DESCRIPTION
**Pull request by snorlaX-sleeps**
#### What this PR does / why we need it:
This PR adds annotations to the Halyard serviceAccount created in this Chart
This is required to enable [IRSA](https://aws.amazon.com/blogs/opensource/introducing-fine-grained-iam-roles-service-accounts/) for pod authentication in AWS

This PR is preliminary work as both Halyard and Spinnaker (CloudDriver / Front-50) need updated to support WebIdentityToken based logins.

This was introduced in `v1.11.623` of the AWS SDK

Added change when annotation is set:
```
---
# Source: spinnaker/templates/rbac/halyard-sa.yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: spinnaker-spinnaker-halyard
  namespace: default
  labels:
    app: "spinnaker-spinnaker"
    heritage: "Helm"
    release: "spinnaker"
    chart: "spinnaker-1.23.3"
  annotations:
    eks.amazonaws.com/role-arn: <role_arn>
```

#### Which issue this PR fixes
  - fixes https://github.com/helm/charts/issues/20885

#### Special notes for your reviewer:
Due to the configuration of Spinnaker via the `-install-using-hal` pod, we can pass CloudDriver and Front-50 `serviceAccountName` values in the `service_settings.yaml` file - this means only the Halyard pod needs an update in the Chart.

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

**Comment by snorlaX-sleeps** [comment]
(This has committed using the wrong account for some reason, apologies)

**Comment by snorlaX-sleeps** [comment]
/assign @ezimanyi
(as requested by robot)

**Comment by aRobinson-R7** [comment]
@zanhsieh any thoughts on this?

**Comment by zanhsieh** [comment]
@viglesiasce
@ezimanyi
@dwardu89
@paulczar
Could you guys LGTM this PR please? Thank you.

**Comment by zanhsieh** [comment]
@snorlaX-sleeps 
You need to bump this chart version to latest in order to pass e2e test.

**Comment by abatilo** [comment]
@snorlax-sleeps

Anything I can do to help push this along?

**Comment by snorlaX-sleeps** [comment]
@AndriMar / @abatilo - I am unsure if this is supported in Spinnaker / Halyard yet, haven't really taken a look since I made this PR

**Comment by snorlaX-sleeps** [comment]
@zanhsieh - not sure why that last test failed
```
I0611 14:15:54.668] ------------------------------------------------------------------------------------------------------------------------
I0611 14:15:54.668] <== Logs of container spinnaker-apx8vnqc76-spi-halyard-0
I0611 14:15:54.669] ------------------------------------------------------------------------------------------------------------------------
I0611 14:15:54.669] ========================================================================================================================
I0611 14:15:54.669] Deleting release 'spinnaker-apx8vnqc76'...
I0611 14:16:15.443] E0611 14:16:15.436475     385 portforward.go:233] lost connection to pod
I0611 14:16:15.444] Error: transport is closing
I0611 14:16:15.444] Error deleting Helm release: Error waiting for process: exit status 1
I0611 14:16:15.444] Deleting namespace 'spinnaker-presubmit-21060-1271083298633814016-apx8vnqc76'...
I0611 14:16:15.722] namespace "spinnaker-presubmit-21060-1271083298633814016-apx8vnqc76" deleted
I0611 14:16:54.957] Namespace 'spinnaker-presubmit-21060-1271083298633814016-apx8vnqc76' terminated.
I0611 14:16:54.957] Error linting and installing charts: Error processing charts
I0611 14:16:54.957] ------------------------------------------------------------------------------------------------------------------------
I0611 14:16:54.958]  ✖︎ spinnaker => (version: "2.0.0-rc5", path: "stable/spinnaker") > Error waiting for process: exit status 1
I0611 14:16:54.958] ------------------------------------------------------------------------------------------------------------------------
I0611 14:16:55.593] f5e76211be6a3a74b1be923013730fd3278b59c2c213fec1d6e9c2ebb92f2599
W0611 14:16:55.614] Traceback (most recent call last):
W0611 14:16:55.614]   File "/workspace/./test-infra/jenkins/../scenarios/execute.py", line 50, in <module>
W0611 14:16:55.614]     main(ARGS.env, ARGS.cmd + ARGS.args)
W0611 14:16:55.615]   File "/workspace/./test-infra/jenkins/../scenarios/execute.py", line 41, in main
W0611 14:16:55.615]     check(*cmd)
W0611 14:16:55.615]   File "/workspace/./test-infra/jenkins/../scenarios/execute.py", line 30, in check
W0611 14:16:55.615]     subprocess.check_call(cmd)
W0611 14:16:55.615]   File "/usr/lib/python2.7/subprocess.py", line 190, in check_call
W0611 14:16:55.615]     raise CalledProcessError(retcode, cmd)
W0611 14:16:55.616] subprocess.CalledProcessError: Command '('./test/e2e.sh',)' returned non-zero exit status 1
```
re-running test

**Comment by zanhsieh** [comment]
@snorlaX-sleeps 
Another PR in Grafana chart has exact the same issue.

/test pull-charts-e2e

**Review by dwardu89 — APPROVED** [review]

File: stable/spinnaker/templates/rbac/halyard-sa.yaml [hunk 1]
```diff
@@ -10,4 +10,8 @@ metadata:
   namespace: {{ .Release.Namespace }}
   labels:
 {{ include "spinnaker.standard-labels" . | indent 4 }}
+  {{- if .Values.serviceAccount.serviceAccountAnnotations }}
+  annotations:
+{{ toYaml .Values.serviceAccount.serviceAccountAnnotations | indent 4 }}
```
Comment (dwardu89) [inline stable/spinnaker/templates/rbac/halyard-sa.yaml:15]: You need to fix the indentation try `nindent 6`

**Review by dwardu89 — CHANGES_REQUESTED** [review]
You need to fix the indentation

File: stable/spinnaker/values.yaml [hunk 2]
```diff
@@ -314,6 +314,7 @@ serviceAccount:
   # If left blank it is auto-generated from the fullname of the release
   halyardName:
   spinnakerName:
+  annotations: {}
```
Comment (dwardu89) [inline stable/spinnaker/values.yaml]: Can you put an example of it as well here please.
```
serviceAccount:
    serviceAccountAnnotations: {}
```

**Comment by AndriMar** [comment]
> @AndriMar / @abatilo - I am unsure if this is supported in Spinnaker / Halyard yet, haven't really taken a look since I made this PR

@snorlaX-sleeps It works with halyard but I have an issue on spinnaker and PR on clouddriver regarding IRSA.
https://github.com/spinnaker/spinnaker/issues/5857

File: stable/spinnaker/templates/rbac/halyard-sa.yaml (hunk 1 above)
Comment (snorlaX-sleeps) [reply stable/spinnaker/templates/rbac/halyard-sa.yaml:15]: I would have thought 4 was correct? an indent of 2 below annotations?

**Comment by snorlaX-sleeps** [comment]
@dwardu89 - tests seem to have passed

**Comment by dwardu89** [comment]
@snorlaX-sleeps You've got a conflict can you update it please?

**Comment by snorlaX-sleeps** [comment]
@dwardu89 updated

**Pull request merged**